### ISSUE TITLE
Sanity check failure during initial reasoning

### ISSUE THREAD

**Issue by sei-eschwartz** [cmu-sei/pharos#139]
@Trass3r is encountering this problem in #137 now that the RTTI facts have been repaired.

**Comment by sei-eschwartz** [comment]
I will look into this as soon as possible, but that may not be until next week.

**Comment by sei-ccohen** [comment]
I've made some progress on this.    I identified two potential issues.   The first is specific to this file.  There's something strange going on with the method at 0x634d60.   It appears in a bunch of unrelated virtual function tables as if was purecall, but it's not the usual kind of purecall for sure (which is why it didn't match any known signatures for purecall).  Instead, it's simply a call to amsg_exit().   Regardless, this method has the _properties_ of being a purecall method, which is to say that we should not be merging classes based on them sharing this method.   To correct this, I added the following fact manually:

purecall(0x634d60).

The other issue appears to be a bug in our Prolog rules.  I want to think about the broader consequences some more before committing the fix publicly, but if you want to proceed, trying removing the "+4" [here](https://github.com/cmu-sei/pharos/blob/master/share/prolog/oorules/rules.pl#L1005).   The problem seems to be that we've not been consistent in deciding whether the GTE (greater than or equal to) size of a VFTable includes the length of the final pointer.  This seems like the kind of situation where we could have lots of code that was confused, and so I should review other uses before committing.  It also possible that this change simply delays the real underlying problem, and that while it's a bug, it's not really _this_ bug per se.  What is fairly obvious to me is that with this +4, we immediately (and possibly incorrectly) conclude that all entries up to the LTE size are valid during the initial reasoning phase and that this triggers the merge conflict, and prevents us from reaching the guessing phase where we're less sensitive to conflicts.

A while back we added the rule to fail immediately after the initial reasoning step if there were any conflicts detected.   This is because our forward rules should be sound enough that this does not occur.   The new behavior is likely to trigger more failures, but it should be very effective at identifying outstanding problems in the forward reasoning rules, so please bear with us while we get those rules to the level of soundness that was intended.

**Comment by sei-ccohen** [comment]
So as I suspected it might, the VFTableSize bug was really just making another problem much more obvious.   Unfortunately, that problem turned out to be #135. :-(   I guess technically you could call the  0x634d60 amsg_exit() issue one of the shared implementations as well.

In the end, by grepping for the addresses that were complained about in the merge conflict message and looking at the MergeClasses rules that caused the merges, I was able to identify this list of functions that were causing problems:

0x40d440 ret
0x5671e0 ret 0x4
0x580470 ret 0x8
0x60f1d0 ret 0xc
0x5671d0 ret 0x10
0x402ad0 xor eax,eax; ret
0x628e30 xor eax,eax; ret 0x4
0x443150 xor eax,eax; ret 0x8
0x517400 xor eax,eax; ret 0xc
0x628b70 xor eax,eax; ret 0x10
0x4b4c20 mov eax,0x1; ret
0x628bc0 mov eax,0x1; ret
0x5173b0 mov eax,0x1; ret 0x8
0x634d60 push 0x19; call amsg_exit; add esp,0x4; ret

Adding facts for "purecall(<address>)." for each of these also prevents the forward reasoning conflict.  This approach may actually give better results than the fix suggested above because it prevents these functions from causing "weird" contradictory merges. For anyone interested in reproducing this process on their executables, this was a little tedious but not too difficult.  I grepped for lines containing both of the conflicting addresses in the log, and then for MergeClasses, then piped the output to awk and got a unique list of methods causing problems in that run.   Then I checked each method by hand to see if it seemed sufficiently "trivial" to have been reused and added a "purecall()" fact as a hack.  I had to repeat the process, running ooprolog.pl 6-7 times to get the complete list.  The culprit was usually reasonMergeClasses_H, which I believe is a sound rule excluding this implementation reuse problem, which technically affects a _lot_ of our rules.

So now, the question is how do we detect this optimization in a fully automated way?

**Comment by Trass3r** [comment]
> I grepped for lines containing both of the conflicting addresses in the log, and then for MergeClasses, then piped the output to awk and got a unique list of methods causing problems in that run.

I guess a shell one-liner might have been easier to type than the explanation ;)

> So now, the question is how do we detect this optimization in a fully automated way?

So this is only a problem if the same function is referenced in 2 different vtables that aren't related or if it's not an override, right?
Cause these return 0 and return 1 functions actually turn up regularly in practice, also reused inside a single vtable.

**Comment by sei-ccohen** [comment]
> I guess a shell one-liner might have been easier to type than the explanation ;)

grep -e <addr1> DKII.EXE.log | grep -e <addr2> | grep reasonMergeClasses | awk '{print $5}' | sort | uniq -c

Not everyone finds shell one liners informative. But I'm happy to oblige.:-)

> So this is only a problem if the same function is referenced in 2 different vtables that aren't related or if it's not an override, right?

That's certainly the common case that's problematic yes.   In general implementation reuse is a case where the source code specification just doesn't match the binary implementation I think. :-(

> Cause these return 0 and return 1 functions actually turn up regularly in practice, also reused inside a single vtable.

Unfortunately under certain compilation options, yes.  Under other compilation options, apparently not at all.  

The problem seems to hinge on the fact that in the source code you can't say things like ClsA::method = ClsB::method or ClsA::vftable[4] = ClsB::method.   You can only say ClsA::method() { return 1; } ClsB::method() { return 1; } or maybe Cls::method() { return ClsB::method(); }

The same code compiled with /DEBUG or even without /LTCG will not exhibit this behavior in our experience.  In both of those latter cases without link-time code generation and similar optimizations you'll get two functions, and eveything is fine in OOAnalyzer.  It's only when the compuiler starts saying "As a really nifty optimization, I can put a completely unrelated method from one class on another class" that you start having the problem. :-(

That's not to say it's not serious problem.  It is.  It's just a difficult one to solve.

**Comment by Trass3r** [comment]
> The same code compiled with /DEBUG or even without /LTCG will not exhibit this behavior in our experience

It is a pure linker optimization for msvc and the default for release builds (though depending on compiler options /Gy and /Gw to work): https://docs.microsoft.com/en-us/cpp/build/reference/opt-optimizations
With gcc the compiler does it already when using -O2/-Os (-fipa-icf): https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html
Hence LTO does make a difference there.

**Comment by Trass3r** [comment]
> > So this is only a problem if the same function is referenced in 2 different vtables that aren't related or if it's not an override, right?
> 
> That's certainly the common case that's problematic yes.

So can that be expressed in Prolog?

**Comment by sei-ccohen** [comment]
We're working on some "heuristic" rules that seem to be working, at least on this sample.   Hopefully they'll be public before too long and you can see them then.

**Comment by Trass3r** [comment]
On a related note, does it check if methods in the same vtable belong to the same or a parent class?
So it could detect classes that haven't been merged yet.

**Comment by sei-eschwartz** [comment]
> On a related note, does it check if methods in the same vtable belong to the same or a parent class?
> So it could detect classes that haven't been merged yet.

I didn't really follow this.

**Comment by Trass3r** [comment]
Well if there are methods in the same final vtable that don't belong to the same or related classes obviously some classes were not merged properly, right?

**Comment by sei-eschwartz** [comment]
> Well if there are methods in the same final vtable that don't belong to the same or related classes obviously some classes were not merged properly, right?

There are several rules in OOAnalyzer that express the fact that methods occurring in the same vftable are related in some way.

But there is a fundamental assumption in OOAnalyzer that you can identify a method at the source-code level by its address in the binary code.  When we see two methods in the same vftable, because of this assumption, we generally assume that they are related in some way, and that the class they are on is related in some way.

But if one of those methods is a shared implementation, then the implementation in the binary corresponds to multiple methods at the source-code level, and these methods may be on classes that are not related at all.  If we don't detect the shared implementation, we can wrongly conclude that these classes are related, and this causes serious problems like the sanity check you encountered.

Another way of looking at this is if you have two methods at the binary level, and one of them is a shared implementation, we can't really answer the question "Are the methods on a related class?"  Because the address doesn't even signify a unique method at the source-code level, so we can't talk about the method being on a single source-code class.

I am not sure if that really answers your question, but when shared implementations are in play it makes it difficult to even express the idea of being on the same class or related classes.

**Comment by Trass3r** [comment]
Ok good so it's also related to this / #135.

**Comment by sei-eschwartz** [comment]
Yes, @sei-ccohen shared a list of functions that he found to be problematic in your binary.  I believe all of these are shared implementations.  We have a new branch that detects all but one of those automatically.  We're working on releasing that but just found a bad regression on one of our test files that we're trying to address first.

**Comment by Trass3r** [comment]
Yes, commonly they even appear multiple times in the same vtable. But I'm sure you already incorporated that heuristic.

**Comment by sei-eschwartz** [comment]
The new branch only tries to identify possible shared implementations by looking for very trivial methods.  Some time ago we talked about trying to automatically identify purecall by looking for methods that appeared many times in vftables.  But we looked into this and there wasn't a good cutoff value.

But I don't think we've really looked at the same method appearing more than once in a single vftable.  If that can't happen normally that would be a great heuristic.  We'll have to look at some of the programs for which we have ground truth to ensure that this doesn't happen for reasons besides shared implementations.

**Comment by Trass3r** [comment]
> But I don't think we've really looked at the same method appearing more than once in a single vftable. If that can't happen normally that would be a great heuristic.

Pure virtuals is the only case I can think of where this happens "naturally" rather than as a subsequent optimization.

**Comment by sei-eschwartz** [comment]
> Pure virtuals is the only case I can think of where this happens "naturally" rather than as a subsequent optimization.

Yup, I thought of that too.  But for similar reasons we don't want to merge because of pure virtuals so that is actually a beneficial side-effect.

**Comment by sei-ccohen** [comment]
Actually, we did look at this already, and it didn't really work out.   Obviously, purecall is the _stereotypical_ reason for a single method to appear repeatedly in the same table.  But it seems that it's also a very common paradigm for reused implementations to appear in the same table multiple times as well.  Imagine having multiple methods on a class that all return a boolean value.  Then you "stub" off some of them with implementations that always just return true or false.  The reused implementation optimization results in the same method being listed several times in the VFTable.

But you were saying "naturally", and you may be correct that without the optimization, it only occurs for purecall.   Another case where we encountered some difficulty was with "reused" deleting destructors. The compiler automatically generates a scalar deleting destructor and a vector deleting destructor, and I'm fairly certain that it's common for both VFTable entries to point to the same implementation.   Whether that's the "optimization" or not becomes a bit philosophical.

In the end, we can't completely ignore the the reused implementation "optimization", so we can't really detect purecall this way.  The new rule that we're trying to push actually uses some logic like this though to help detect reused implementations (including purecall).

**Comment by sei-eschwartz** [comment]
@sei-ccohen Why do we have to distinguish between purecall and shared implementations?  In both cases we don't want to merge them.

**Comment by Trass3r** [comment]
Exactly, detected purecall signatures is just another heuristic to find shared implementations :)

**Comment by sei-ccohen** [comment]
This issue is largely addressed by the recent commit, but the amsg_exit() method at 0x634d60 is still a problem.   I think this should run to completion now with only purecall(0x634d60) appended to the facts file.  We still don't have an approach for non-trivial reused implementations.  Please let us know and close this ticket if you're able to confirm.

**Comment by Trass3r** [comment]
Btw the parameters I'm using are `--serialize ser.dat -F facts.pl --timing --per-function-timeout 4000 --maximum-memory 30000 --threads=8 -n 634E80 --delete-method 6341B0 -n 57CEB0 --delete-method 57D0B0`.

**Comment by sei-eschwartz** [comment]
It is normal for consistency checks to fail after guesses.  I am not sure why those classes would lose the `vftptr_0` members though.  They clearly have a vftable.

**Comment by sei-eschwartz** [comment]
`vftptr_0` is definitely missing from 0x66db54 and should not be.

**Comment by sei-eschwartz** [comment]
There is an entry in the `.facts` file `methodMemberAccess(0x4d255f, 0x4d2550, 0, 0x4).`

**Comment by sei-eschwartz** [comment]
Oops, I think this is a side-effect of using `purecall`.  The class containing `purecall` becomes worthless, which triggers an error in `oojson` such that the `vftptr` member is not printed.

The better workaround would be to put `trivial(0x1234)` somewhere like `rules.pl`.

**Comment by sei-eschwartz** [comment]
Just to conclude, I think the only issue still remaining here is that we don't detect the call to `amsg_exit()` as a shared implementation.

**Comment by Trass3r** [comment]
> Oops, I think this is a side-effect of using `purecall`. The class containing `purecall` becomes worthless, which triggers an error in `oojson` such that the `vftptr` member is not printed.
> 
> The better workaround would be to put `trivial(0x1234)` somewhere like `rules.pl`.

`trivial(0x634d60).` doesn't work: `Invalid predicate: trivial/1`

**Comment by sei-eschwartz** [comment]
> `trivial(0x634d60).` doesn't work: `Invalid predicate: trivial/1`

It looks like trivial is defined in `initial.pl` instead of `rules.pl` like I suggested.

**Comment by Trass3r** [comment]
(I added that line to the generated facts file.)

**Comment by sei-eschwartz** [comment]
I think that should be fine, although you must have turned off some of the checks for the facts file.

**Comment by Trass3r** [comment]
Not sure what you mean, I just used ooprolog: https://github.com/cmu-sei/pharos/blob/master/share/prolog/oorules/README.md#step-three-prolog-analysis

**Comment by sei-eschwartz** [comment]
I see, you got that error from putting trivial in the fact file.  Put it in `initial.pl` instead.

**Comment by Trass3r** [comment]
Feels rather hacky to modify the pharos files.
Anyway with _and_ without the `purecall(0x634d60).` hack I now get 
```
tryBinarySearch completely failed on [(0x674290, 0x62a180)] and will now backtrack to fix an upstream problem.
Refusing to backtrack into reasoningLoop to fix an upstream problem because backtrackForUpstream/0 is not set.
This likely indicates that there is a problem with the OO rules.
```
with the latest docker image.

**Comment by sei-eschwartz** [comment]
Just to make sure I understand, with `purecall(0x634d60).` you used to be able to run to completion, and now you get the above error?

**Comment by Trass3r** [comment]
IIRC yes. It's become hard to follow all these tickets and which are related to the shared implementation issue.

**Comment by sei-eschwartz** [comment]
I'm going to open a new issue for this, because I don't think it is (necessarily) related to the original problem.

**Comment by sei-eschwartz** [comment]
I think this is resolved.  Please re-open if not.
